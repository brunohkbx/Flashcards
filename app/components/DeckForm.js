import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { HelperText, TextInput } from 'react-native-paper';
import { View } from 'react-native';

export const FormikDeckForm = ({ handleSubmit, initialValues, formRef }) => {
  const validationSchema = Yup.object().shape({
    title: Yup.string().required('Required')
  });

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      enableReinitialize
      validateOnBlur={false}
      ref={formRef}
      component={DeckForm}
      onSubmit={(values, actions) => handleSubmit(values, actions)}
    />
  );
};

export const DeckForm = props => {
  const {
    handleChange,
    handleBlur,
    touched,
    errors,
    values,
  } = props;

  return (
    <View>
      <TextInput
        label="Title"
        value={values.title}
        onChangeText={handleChange('title')}
        onBlur={handleBlur('title')}
        error={!!(touched.title && errors.title)}
        autoFocus
      />
      <HelperText type="error" visible={!!(touched.title && errors.title)}>
        {errors.title}
      </HelperText>
    </View>
  );
};

FormikDeckForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.object,
  formRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
};

FormikDeckForm.defaultProps = {
  initialValues: { title: '', questions: [] }
};

export default FormikDeckForm;
