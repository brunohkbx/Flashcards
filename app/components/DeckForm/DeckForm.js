import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Formik, FieldArray } from 'formik';
import * as Yup from 'yup';
import uuid from 'uuid';
import { View, KeyboardAvoidingView } from 'react-native';
import FAB from '../FAB';
import BottomRightContainer from '../BottomRightContainer';
import { Header } from 'react-navigation';
import FormContent from './FormContent';

export const FormikDeckForm = ({ handleSubmit, initialValues, formRef }) => {
  const validationSchema = Yup.object().shape({
    title: Yup.string().required('Required'),
    questions: Yup.array().of(
      Yup.object().shape({
        question: Yup.string().required('Required'),
        answer: Yup.string().required('Required'),
        id: Yup.string().required('Required')
      })
    )
  });

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      enableReinitialize
      validateOnBlur={false}
      ref={formRef}
      render={formikProps => (
        <FieldArray name="questions" component={DeckForm} />
      )}
      onSubmit={(values, actions) => handleSubmit(values, actions)}
    />
  );
};

export class DeckForm extends Component {
  addNewFlashcard = () => {
    const { push } = this.props;

    push({ question: '', answer: '', id: uuid() });
  }

  handleFlashcardDeleted = index => {
    const { remove } = this.props;

    remove(index);
  }

  render() {
    return (
      <View style={{flex: 1, padding: 8}}>
        <KeyboardAvoidingView
          style={{flex: 1}}
          behavior="padding"
          enabled
          keyboardVerticalOffset={Header.HEIGHT + 35}
        >
          <FormContent
            onFlashcardDeleted={this.handleFlashcardDeleted}
            {...this.props.form}
          />
        </KeyboardAvoidingView>
        <BottomRightContainer right={16} bottom={16} >
          <FAB handlePress={this.addNewFlashcard} />
        </BottomRightContainer>
      </View>
    );
  }
}

FormikDeckForm.propTypes = {
  initialValues: PropTypes.shape({
    title: PropTypes.string.isRequired,
    questions: PropTypes.array.isRequired,
    id: PropTypes.string.isRequired
  }),
  formRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  handleSubmit: PropTypes.func.isRequired,
};

export default FormikDeckForm;
