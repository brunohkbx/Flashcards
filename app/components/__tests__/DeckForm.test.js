import React from 'react';
import { shallow } from 'enzyme';
import { FormikDeckForm, DeckForm } from '../DeckForm';

describe('FormikDeckForm', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      handleSubmit: jest.fn(),
    }, propOverrides)

    const wrapper = shallow(<FormikDeckForm {...defaultProps} />);

    const submit = () => wrapper.simulate('submit');

    return {
      wrapper,
      submit
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  test('onSubmit calls handleSubmit', () => {
    const mockHandleSubmit = jest.fn();
    const { submit } = setup({ handleSubmit: mockHandleSubmit });

    submit();

    expect(mockHandleSubmit).toHaveBeenCalled();
  });
});

describe('DeckForm', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      handleChange: jest.fn(() => jest.fn),
      handleBlur: jest.fn(() => jest.fn),
      touched: {},
      errors: {},
      values: {},
    }, propOverrides)

    const wrapper = shallow(<DeckForm {...defaultProps} />);

    return {
      wrapper
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('displays an error message when the input is invalid', () => {
    const { wrapper } = setup({
      touched: { title: true },
      errors: { title: 'Required' }
    });

    expect(wrapper).toMatchSnapshot();
  });
})
