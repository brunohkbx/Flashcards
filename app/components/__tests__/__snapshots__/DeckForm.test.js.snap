// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckForm displays an error message when the input is invalid 1`] = `
<Component>
  <withTheme(TextInput)
    autoFocus={true}
    error={true}
    label="Title"
    onBlur={[Function]}
    onChangeText={[Function]}
  />
  <withTheme(HelperText)
    type="error"
    visible={true}
  >
    Required
  </withTheme(HelperText)>
</Component>
`;

exports[`DeckForm renders properly 1`] = `
<Component>
  <withTheme(TextInput)
    autoFocus={true}
    error={false}
    label="Title"
    onBlur={[Function]}
    onChangeText={[Function]}
  />
  <withTheme(HelperText)
    type="error"
    visible={false}
  />
</Component>
`;

exports[`FormikDeckForm renders properly 1`] = `
<Formik
  component={[Function]}
  enableReinitialize={true}
  initialValues={
    Object {
      "questions": Array [],
      "title": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  validateOnBlur={false}
  validateOnChange={true}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": false,
      "_nodes": Array [
        "title",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "title": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
/>
`;
